security/sensorhub-security-oauth/src/main/java/org/sensorhub/impl/security/oauth/OAuthAuthenticator.java

```diff
@@ -227,23 +231,73 @@ public Authentication validateRequest(ServletRequest req, ServletResponse resp,
                 }
             }
 
-            if (authHeader != null && !request.getServletPath().equals("/admin")) {
+            // if no OAuth token received check if credentials are provided in authorization header
+            // and use back channel direct auth flow
+            if (accessToken == null && authHeader != null && !request.getServletPath().equals("/admin")) {
+                try {
+                    var credentials = parseBasicAuth(authHeader);
+                    if (credentials == null) {
+                        var clientCredentialsToken = parseBearerToken(authHeader);
+
+                        if (clientCredentialsToken != null) {
+                            // if client credentials token, assume it's a JWT token that contains needed info
+                            log.debug("Client Credentials Token = {}", clientCredentialsToken);
+
+                            String[] chunks = clientCredentialsToken.split("\\.");
+
+                            Base64.Decoder decoder = Base64.getUrlDecoder();
+
+                            String header = new String(decoder.decode(chunks[0]));
+
+                            log.debug("Client Credentials Token Header = {}", header);
+
+                            JsonElement headerData = JsonParser.parseString(header);
 
-                clientCredentialsToken = parseBearerToken(authHeader);
+                            JwkProvider jwkProvider =
```

Comment:
I think to benefit from the cache this would have to be created elsewhere, right?
Here it looks like it's recreated on every auth request